testing/matrix-sdk-integration-testing/src/tests/room_privacy.rs

```diff
@@ -0,0 +1,231 @@
+
+#[tokio::test]
+async fn test_publishing_room_alias() -> anyhow::Result<()> {
+    let client = TestClientBuilder::new("alice").build().await?;
+    let server_name = client.user_id().expect("A user id should exist").server_name();
+
+    let sync_handle = tokio::task::spawn({
+        let client = client.clone();
+        async move {
+            client.sync(SyncSettings::default()).await.expect("Sync should not fail");
+        }
+    });
+
+    // The room can only be visible in the public room directory later if its join
+    // rule is one of  [public, knock, knock_restricted] or its history
+    // visibility is `world_readable`. Let's use this last option.
+    let room_history_visibility = InitialRoomHistoryVisibilityEvent::new(
+        RoomHistoryVisibilityEventContent::new(HistoryVisibility::WorldReadable),
+    )
+    .to_raw_any();
+    let room_id = client
+        .create_room(assign!(CreateRoomRequest::new(), {
+            initial_state: vec![
+                room_history_visibility,
+            ]
+        }))
+        .await?
+        .room_id()
+        .to_owned();
+
+    // Wait for the room to be synced
+    let room = client.await_room_remote_echo(&room_id).await;
+
+    // Initial checks for the room's state
+    let room_visibility = room.privacy_settings().get_room_visibility().await?;
+    assert_matches!(room_visibility, Visibility::Private);
+
+    let canonical_alias = room.canonical_alias();
+    assert!(canonical_alias.is_none());
+
+    let alternative_aliases = room.alt_aliases();
+    assert!(alternative_aliases.is_empty());
+
+    // We'll add a room alias to it
+    let random_id: u128 = random();
+    let raw_room_alias = format!("#a-room-alias-{random_id}:{server_name}");
+    let room_alias = RoomAliasId::parse(raw_room_alias).expect("The room alias should be valid");
+
+    // We publish the room alias
+    let published =
+        room.privacy_settings().publish_room_alias_in_room_directory(&room_alias).await?;
+    assert!(published);
+
+    // We can't publish it again
+    let published =
+        room.privacy_settings().publish_room_alias_in_room_directory(&room_alias).await?;
+    assert!(!published);
+
+    // We can publish an alternative alias too
+    let alt_alias = RoomAliasId::parse(format!("#alt-alias-{random_id}:{server_name}"))
+        .expect("The alt room alias should be valid");
+    let published =
+        room.privacy_settings().publish_room_alias_in_room_directory(&alt_alias).await?;
+    assert!(published);
+
+    // Since we only published the room alias, the canonical alias is not set yet
+    let canonical_alias = room.canonical_alias();
+    assert!(canonical_alias.is_none());
+
+    // We update the canonical alias now
+    room.privacy_settings()
+        .update_canonical_alias(Some(room_alias.clone()), vec![alt_alias.clone()])
+        .await?;
+
+    // Give some time for the sync to run
+    tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
```

Comment:
I think this will be flaky, and if the sync isn't slow it it slows down the test unnecessarily. Can't we listen for the event to arrive using an event handler and use the [timeout](https://docs.rs/tokio/latest/tokio/time/fn.timeout.html) method to make sure that we're not blocking forever if the event never arrives.

Reply:
https://github.com/matrix-org/matrix-rust-sdk/pull/4401/commits/c372fcbf18544a2e565576f94401d27c2625d763: I tried using Mutex but I couldn't get it to work, so I followed a different example which used channels and it worked fine. I wouldn't mind learning how this should be done with Mutexes though 😅 .